### PR TITLE
Zika working for modular augur

### PR DESCRIPTION
**Pull request by jameshadfield**
Hi @emmahodcroft @rneher @trvrb 

Zika is now building using modular augur via `snakemake auspice/zika_tree.json` (see https://nextstrain.org/docs/builds/zika-build) and is up on the staging server. Subsampling has been turned off, although perhaps this isn't what's wanted for the live builds (simply change the value of `vpc` in the snakemake `filter` rule). Instructions on installing / running modular augur have been started here: https://nextstrain.org/docs/bioinformatics-pipeline/modular-augur

## General changes to modular augur:
* The alignment script is slightly changed to fix a bug where we would get large blocks of `N`s for all sequences (including the reference!)
* if `vpc=0` then no subsampling is performed
* The positions of nuc & aa mutations are incremented by 1 to match the live nextstrain.org/zika naming. Does this work for TB?
* The arguments to `augur export` have been modified to accept an arbitrary number of node-decorating JSON files, and the names have been changed (e.g. `output` is now `tree_output`). **This means all snakemake files will need to be updated**
* `node.clade` is deprecated (but remains in the JSONs and is now equal `node.attr.strain` to prevent auspice errors)
* Some non-essential fields have been removed from the metadata JSON, but their generating functions remain in the codebase in case nextflu requires them

## Snakemake file structure
Following Emma's advice i've defined the filepaths at the top of the (zika) snakemake file, but left paramaters etc in their respective rules.

### CODE REVIEW

File: augur/export.py [hunk 1]
```diff
@@ -27,6 +29,10 @@ def tree_to_json(node, fields_to_export = [], top_level = [], div=0):
             if hasattr(node, fname):
                 val = node.__getattribute__(fname)
 
+        # shadow clade by strain. clade is deprecated and will be removed.
+        if field == "clade"
```
Comment (huddlej) [inline augur/export.py]: Missing a colon here. The "clade" attribute is only deprecated for auspice, but not for augur, right?

File: augur/filter.py [hunk 2]
```diff
@@ -82,6 +82,8 @@ def run(args):
 
     if args.cat and args.viruses_per_cat:
         vpc = args.viruses_per_cat
+        if vpc == 0:
```
Comment (huddlej) [inline augur/filter.py]: This doesn't seem necessary. The default value for `viruses_per_cat` is `None`, so the user can skip subsampling by omitting the `--viruses_per_cat` flag in the Snakefile. If they specify a value of 0, the preceding if statement should evaluate to false and use the complete data set.

File: builds/zika/Snakefile [hunk 3]
```diff
@@ -1,24 +1,40 @@
+# config variables to be used by rules below
```
Comment (huddlej) [inline builds/zika/Snakefile]: We should include an `all` rule at the top of each Snakefile that includes one or more final outputs to be built. This special rule allows users to simply type `snakemake` from the proper build directory without needing to specify the output filename (e.g., `zika_tree.json`).

**Comment by tsibley** [comment]
> Some non-essential fields have been removed from the metadata JSON, but their generating functions remain in the codebase in case nextflu requires them

Can you document these somewhere so we can know what to cleanup later without going digging?

**Comment by tsibley** [comment]
> This `tmp` variable is only undefined when the following logic evaluates to false: `if c.name in aln and n.name in aln` on line 289.

The logic in that stanza is similar to `assign_aa_vcf()` in that same file, which contains a `tmp = []` initialization as the first statement in the loop `for c in n:`.  I suspect the same initialization wants to happen in the problematic stanza.

**Comment by tsibley** [comment]
…and looking closer, I suspect that the whole stanza with `tmp`:

```python
    ## determine amino acid mutations for each node
    if is_vcf:
        aa_muts = assign_aa_vcf(tree, translations)
    else:
        aa_muts = {}
        for n in tree.get_nonterminals():
            for c in n:
                aa_muts[c.name]={"aa_muts":{}}
            for fname, aln in translations.items():
                for c in n:
                    if c.name in aln and n.name in aln:
                        tmp = [(a,pos,d) for pos, (a,d) in
                                enumerate(zip(aln[n.name], aln[c.name])) if a!=d]
                    aa_muts[c.name]["aa_muts"][fname] = tmp
```

should be refactored into an `assign_aa()` function that parallels `assign_aa_vcf()`:

```python
def assign_aa_vcf(tree, translations):
    aa_muts = {}

    for n in tree.get_nonterminals():
        for c in n:
            aa_muts[c.name]={"aa_muts":{}}
        for fname, prot in translations.items():
            n_muts = prot['sequences'][n.name]
            for c in n:
                tmp = []
                c_muts = prot['sequences'][c.name]
                for pos in prot['positions']:
                    #if pos in both, check if same
                    if pos in n_muts and pos in c_muts:
                        if n_muts[pos] != c_muts[pos]:
                            tmp.append((n_muts[pos],int(pos+1),c_muts[pos]))
                    elif pos in n_muts:
                        tmp.append((n_muts[pos],int(pos+1),prot['reference'][pos]))
                    elif pos in c_muts:
                        tmp.append((prot['reference'][pos],int(pos+1),c_muts[pos]))

                aa_muts[c.name]["aa_muts"][fname] = tmp
```

I'm not quite sure that they can't be the same function either, as I don't immediately follow the exact logic in the innermost loops.  (It would be nice to not use meaningless variable names like `n` and `c` and `tmp`.)

File: augur/align.py [hunk 4]
```diff
@@ -62,13 +62,13 @@ def run(args):
...
         seq.seq = seq.seq.upper()
-    if args.fill_gaps:
-        make_gaps_ambiguous(aln)
 
     AlignIO.write(aln, output, 'fasta')
 
     if ref_name:
         seqs = strip_non_reference(output, ref_name, keep_reference=not args.remove_reference)
+        if args.fill_gaps:
+            make_gaps_ambiguous(seqs)
```
Comment (tsibley) [inline augur/align.py]: Should this stanza be guarded by the `if seqs:` conditional immediately below it?

**Comment by jameshadfield** [comment]
>Would you document these somewhere so we can know what to cleanup later without going digging?
* https://github.com/nextstrain/augur/blob/zika/augur/export.py#L95
* https://github.com/nextstrain/augur/blob/zika/augur/export.py#L166

While it seems like there are some niggling code issues (and a pesky alignment issue that's really bizarre), are there any issues with how the `augur ...` scripts are called?

**Comment by tsibley** [comment]
>  are there any issues with how the `augur ...` scripts are called?

I have a couple suggestions:

1. Things like `--viruses_per_cat` are traditionally spelled `--viruses-per-cat`.  There are many examples of underscores being used in long option names, which goes against convention.  I think we should adjust all of those.

2. I suggest not shortening long option names like `--meta_output` or `--cat` or `--viruses_per_cat` but instead spelling them out as `--metadata-output` and `--category` and `--viruses-per-category`.  This makes them exceedingly clear and consistent with similar option names like `--metadata`.

Other than that, I think it's a fine interface.  It can always evolve later if we want to be able to do things like pipe data from stdin and write it to stdout.

The only potential pitfall I see is that providing multi-value options as

    --node_data results/node_data.json results/traits.json results/aa_muts.json

instead of as

    --node_data results/node_data.json --node_data results/traits.json --node_data results/aa_muts.json

can make it a little trickier to add positional arguments later.  I'm not sure we'll ever want to add positional arguments though, so it's nicer to do things the current way.

**Comment by emmahodcroft** [comment]
Also, it would be good to rename that option from `--viruses_per_cat` to `--samples_per_cat` ;)

**Comment by tsibley** [comment]
James, would you drop these two commits from the branch since together they're a no-op?

* [remove all Ns before alignment](https://github.com/nextstrain/augur/pull/128/commits/72ff689f57b796d977f6f0924c15e73686d2ed55)
* [Revert "remove all Ns before alignment"](https://github.com/nextstrain/augur/pull/128/commits/99ec81cdf8815b03bbeb79d030866ddbe2cb8b56)

**Comment by rneher** [comment]
* fine with `--one-two-three` flags and spelling things out
 * +1 for `--samples-per-category` (or maybe `--sequences-per-category`)
 * the translation functions can't be exactly the same as the vcf version stores sequences in different ways (1e7 vs 1e4 long). But the specific parts could be moved further inside the loop and made two simple functions

**Comment by tsibley** [comment]
+1 for `--sequences-per-category` instead of `--viruses_per_cat`.  It's good to be precise about what's being filtered (sequences), and of course it's not just viruses anymore as Emma pointed out. :-)

**Comment by jameshadfield** [comment]
re: `this-then-that` parameters, a quick test makes me think one cannot write this in a Snakemake file:
```
	params:
		n-iqd = 4
	shell:
		'... --n-iqd {params.n-iqd}'
```

**Comment by tsibley** [comment]
> I have the same problem as @huddlej with the -fast flag in IQTree from installing through bioconda. If we could add a temporary parameter to call without the flag until we figure out to handle it better, it would be nice. Alternatively, one could see if there was a way to detect if the call failed because of the -fast flag and automatically try to call again without the flag.

The `-fast` flag is an alias that just bundles other options, which we can instead set directly, bypassing this issue.  I'll push a commit up.

**Comment by tsibley** [comment]
@emmahodcroft @huddlej Give the `zika` branch another shot with your IQ-TREE issues after pulling down this new commit on it? https://github.com/nextstrain/augur/pull/128/commits/6427495ca538c6dddf1d64bccf31e92d4c08b40e

**Comment by tsibley** [comment]
> re: this-then-that parameters, a quick test makes me think one cannot write this in a Snakemake file:

@jameshadfield: Correct; those param assignments follow Python parsing rules.  I was talking about the actual command line parameters, not the parameter names in the Snakefile.  For example:

```
rule …:
    params:
        n_iqd = 4
    shell:
        '… --n-iqd {params.n_iqd}'
```

**Comment by emmahodcroft** [comment]
@tsibley I'll check as soon as I can. I'm about to start travelling and won't have great internet until I'm back in Basel next Monday (but I can definitely check before that - but responses in general are going to be slow!)

**Comment by huddlej** [comment]
@tsibley The explicit iqtree options work great for me. Nice catch!

@emmahodcroft I like the idea of naming outputs at the top of the Snakefile instead of the top and the bottom. One strange convention of Make-style files (from the GNU Make days) is that the files usually read from bottom to top so the first file created is at the bottom and the last file is at the top. This is definitely counter-intuitive, but it is why most of these make tools will run the first rule as the primary output.

**Comment by emmahodcroft** [comment]
@tsibley Yes, the iqtree options work for me too - thanks!

**Comment by tsibley** [comment]
Thanks for testing it, @emmahodcroft and @huddlej!  Glad it works.

**Comment by tsibley** [comment]
> I like the idea of naming outputs at the top of the Snakefile instead of the top and the bottom.

Ditto.

**Comment by jameshadfield** [comment]
re: `this-then-that` params: yup, I know you're talking about the actual command line parameters, i'm just pointing out that these will then diverge from how they're stored in the snakemake file (`this_then_that`). 

I.e. we'll have Snake files like this:
```
rule export:
	input:
		node_data = rules.timetree.output.node_data,
		geo_info = config.geo_info
	shell:
		'augur export  --node-data {input.node_data} --geo-info {input.geo_info}...'

```

**Comment by jameshadfield** [comment]
to do list:
* make arguments more verbose and change to `this-then-that`
* naming outputs at the top of the Snakefile instead of the top and the bottom (john / emma)
* nucleotide & aa +1 fix (james)
anything else?

**Comment by tsibley** [comment]
> i'm just pointing out that these will then diverge from how they're stored in the snakemake file

Ah, gotcha. Sure. I think consistency with Unix conventions (kebab-case for options) is more important than superficial consistency in the Snakefile.

> anything else?

Did you see https://github.com/nextstrain/augur/pull/128#discussion_r195262864?  It may be correct, it just isn't apparent to me, so I wanted to call attention to it.

Would you drop these two commits from the branch since they're a no-op?  If they stay they'll clutter history.

* [remove all Ns before alignment](https://github.com/nextstrain/augur/pull/128/commits/72ff689f57b796d977f6f0924c15e73686d2ed55)
* [Revert "remove all Ns before alignment"](https://github.com/nextstrain/augur/pull/128/commits/99ec81cdf8815b03bbeb79d030866ddbe2cb8b56)

**Comment by jameshadfield** [comment]
Great, thanks tom! I've updated the to-do list

Re: dropping commits. I'm not going to change history if there's a chance people have this branch checked out, but it's fine with me if whoever rebases it wants to clean things up.

**Comment by tsibley** [comment]
> Re: dropping commits. I'm not going to change history if there's a chance people have this branch checked out, but it's fine with me if whoever merges it into master wants to rebase first & clean things up.

Ok, though I think this is a misguided caution for a feature branch and increases risk of a bad merge. When you and Trevor are back and we can get the whole group together, I'd appreciate having a discussion about version control and code review practices so I can better understand what you have in place and what your goals are for them.

**Comment by trvrb** [comment]
@tsibley --- We haven't squashed commits in the past.

@jameshadfield --- Small to clean up:

1. Some places of the `Snakefile` refer to `meta` and others refer to `metadata`. I suggest we standardize to `metadata` (matching module command line interface).
2. Same thing for `seq` vs `sequences` and `aln` vs `alignment`. I would suggest to standardize on `sequences` and `alignment`.
3. I second `sequences_per_category` (don't like mixing `samples` and `sequences` in the interface).
4. I would rename `parse --cat` to `parse --categories`. `cat` makes me think `concat`.

Everyone --- Larger questions:

1. Strange that `augur parse` produces `--metadata` as an output, while `augur filter` takes `metadata` as an input. A similar issue appears in `augur export` where we have `--metadata` as input and `--meta_output` as output. `augur timetree` is also quite confusing with `--tree` as input and a tree as `--output`. There doesn't seem to be an obvious solution, but I would suggest to prepend `output_` to all outputs. So `augur parse` has `--output_sequences` and `--output_metadata` and `augur timetree` has `--output_node_data` and `--output_tree`.

_Need to run. More later..._

**Comment by tsibley** [comment]
> We haven't squashed commits in the past.

Just to be clear, I'm talking about tidying up development commits by dropping them in rebase. I'm not advocating for use of [`git merge --squash`](https://git-scm.com/docs/git-merge#git-merge---squash).

**Comment by huddlej** [comment]
@trvrb 

> 2. I don’t fully understand what's going on with config vs params. You could say that all files should be in config to make Snakemake do it's thing, but then we'd want to remove fasta_fields. Also, we have "parameters" embedded in many of the command line calls like --time_marginal, --date_confidence and --fill_gaps.

The `config` dictionary is for global configuration of anything in the Snakefile while `params` are for rule-specific parameters. It is common to see `config` used to define paths to files required by Snakemake rules when there isn’t a rule to create those files (like the color maps). The `config` dictionary is also useful for any constants or magic numbers that are used in rules. I think the idea behind having a `config.json` file and a `Snakefile` is separation of concerns so the users can drop in a valid configuration file without having to change any rules (e.g., `snakemake --configfile new_config.json`). The current implementation here with a `config` rule and nested `params` is not something I had seen before.

The `params` are mostly helpful because they can be dynamically defined by functions based on wildcard values. Although you can define rule-specific constants and magic numbers with `params`, I’ve usually seen people define these in the `config` dictionary and then reference that dictionary value in the shell rule like so.

```
configfile: "config.json"
config["cores"] = 4

rule map_sequences:
    input: "reads.fasta", "reference.fasta"
    output: "alignments.bam"
    shell: "aligner --cores {config[cores]} {input} > {output}"
```

If you need some sort of conditional logic based on the wildcards, `params` are more powerful. For example, you can [reference a callback function](https://github.com/nextstrain/janus/blob/master/Snakefile#L99) and then [that function](https://github.com/nextstrain/janus/blob/master/Snakefile#L44-L48) gets the values of the wildcards for a given rule instance and can generate parameters dynamically. In that janus example, I actually treated command line flags like `--no_tree_freqs` as parameters. So, if we decided we wanted users to be able to configure whether to run with `--time_marginal` or not, we could make those parameters.

**Comment by rneher** [comment]
I would put less emphasis on the snake files as of now. what we need to have stable are the augur command lines parameters. I agree that we should follow the convention of `-` instead of `_` even if it comes with the pain of not being able to use the arguments as attributes in python code (since `-` is an arithmetic operator). Also ok with spelled out names. `augur parse` is really just a little prepend to unravel the fauna fasta into a sequence and meta file.

**Comment by tsibley** [comment]
> I agree that we should follow the convention of `-` instead of `_` even if it comes with the pain of not being able to use the arguments as attributes in python code

`argparse` helpfully converts `-` to `_` when making the attribute names, so `--foo-bar` gets turned into `args.foo_bar` without any extra work.

**Comment by jameshadfield** [comment]
Update:
* nt & aa position fixes have been moved to the appropriate functions, and should now work with VCF inputs
* alignment module now exports intermediate results (crucial for checking validity of builds)
* named outputs are now at the top of the snakefile (thanks @huddlej)
* arguments are now written `this-then-that`. Very handily, argparse converts these to `this_then_that` for use within the modules themselves. That is, the `augur align` module requires `--fill-gaps`, but inside the `align.py` script, you can refer to `args.fill_gaps` (so nothing outside of the snakemake file(s) and `bin/augur` needed changing :tada:)

**to do:**
@tsibley and @trvrb will change the wording of the arguments as applicable

File: builds/zika/Snakefile [hunk 5]
```diff
@@ -9,118 +9,119 @@ rule all:
...
-			"--viruses-per-cat {params.vpc} "
-			"--exclude {input.exclude} --cat {params.cat} --min-date {params.min_date}"
+    input:
+        sequences = rules.parse.output.sequences,
+        metadata = rules.parse.output.metadata,
+        exclude = config.dropped_strains
+    output:
+        "results/filtered.fasta"
+    params:
+        vpc = 20,
```
Comment (tsibley) [inline builds/zika/Snakefile]: @trvrb This otherwise whitespace-only commit also changed zika's `vpc` parameter from 1 to 20. Was that intended?

Comment (trvrb) [reply builds/zika/Snakefile]: Yes. It should have been 20 all along. See here: https://github.com/nextstrain/augur/blob/master/builds/zika/zika.prepare.py#L16. The "1" was the mistake.

**Comment by trvrb** [comment]
Okay. That's my pass. I'm happy for this to be merged now. I only have one remaining issue regarding the use of `node_data` and the `timetree` module. However, this is a big enough question (and open enough) that I'd prefer to move it into a separate issue.

@tsibley --- Can you take a look now and see if there are any remaining argument wording changes you'd like?

**Comment by trvrb** [comment]
@rneher --- How are ancestral sequences / mutations getting added to `node_data.json` by https://github.com/nextstrain/augur/blob/zika/builds/zika/Snakefile#L73 when `--ancestral` is not called?

**Comment by tsibley** [comment]
@trvrb I wondered about that a bit ago when looking at what the `--timetree` option did, and I think it's because `augur treetime` has three modes: `--timetree`, `--ancestral`, and neither (purely naming internal nodes). I believe that the `--timetree` mode includes the ancestral reconstruction.

**Comment by tsibley** [comment]
Ok, I've reviewed @trvrb's changes to option names and general cleanups.
Overall much happier with the consistent interface! I pushed up what I think
are clear improvements but also have a couple thoughts below that warrant
discussion:

* filter: `--categories` and `--sequences-per-category` would read more
  naturally to me as `--group-by` and `--sequences-per-group`. This aligns with
  standard aggregation/collation terminology, e.g. pandas `.groupby()`, SQL's
  `GROUP BY`.

* all: I'm not sure it's worth an explicit short alias `-s` for `--sequences`.
  `--seq` will work as an abbreviation and it's nice to encourage clarity
  through long option names.

* all: We had talked about making all single-output commands still use
  `--output-<noun>` instead of just `--output`.  Thanks to unambiguous prefix
  matching, this would still allow `--output` to work, but it would maintain
  consistency with multi-output commands and be explicit about what's being
  output.

* translate: It seems like the options `--output`, `--alignment-output`, and
  `--vcf-output` want reconciliation, but I'm not sure what's best.

* translate: `--vcf-input` may want renaming as we've avoided using "input" in
  other option names.

**Comment by trvrb** [comment]
Thanks @tsibley. These are all definite improvement. For specific questions:

>--group-by / --sequences-per-group vs --categories / --sequences-per-category

I'm equally happy either way.

> I'm not sure it's worth an explicit short alias -s for --sequences. --seq will work as an abbreviation and it's nice to encourage clarity through long option names.

Agree in theory. Though it was news to me that you could auto shorten argparse commands.

> We had talked about making all single-output commands still use --output-<noun> instead of just --output. Thanks to unambiguous prefix matching, this would still allow --output to work, but it would maintain consistency with multi-output commands and be explicit about what's being output.

I actually started writing this out in the Snakefile (`--output-sequences`, etc...) and wasn't happy with the seemingly unnecessary verbosity. And felt natural to me to distinguish single output modules from multi output modules.

**Comment by emmahodcroft** [comment]
@tsibley 

> - translate: --vcf-input may want renaming as we've avoided using "input" in
other option names.

We could combine `--node-data` and `--vcf-input` into something like `--ancestral-sequences` (if that doesn't imply it contains _only_ ancestral sequences) and disambiguate by the file ending, but I'm not sure if that might make it more confusing to convey to users which to use - as VCF input users will also have a `node_data.json`, but it doesn't have sequences in it. 
...On the other hand, maybe this would be more clear than what's currently there, as it could be misinterpreted that VCF users should supply both files? (I had to double-check this myself...)

Changing to writing ancestral sequences to a fasta instead of to the json would make this less ambiguous, but would mean one more intermediate output file, so unsure if that's worth it.

> - translate: It seems like the options --output, --alignment-output, and
--vcf-output want reconciliation, but I'm not sure what's best.

We could do something similar to the above for `--alignment-output` and `--vcf-output`, too - just have `--alignment-output` and write out fasta/vcf depending on input. We'd need to make clear that the GENE thing won't work for VCF, though (but worst case is they end up with `translations_GENE.vcf`)

**Comment by trvrb** [comment]
>Is this an "I don't care one way or the other" response, or a "Let's leave it as-is" response?

I like giving explicit short forms at least as an option. So I'd like:

```
augur filter --sequences zika.fasta
augur filter --seq zika.fasta
augur filter -s zika.fasta
```

to all work. Obviously, we should be careful with `-s` and make sure it only means `--sequences` throughout the different augur modules.

**Comment by tsibley** [comment]
@trvrb Got it. Leaving it as-is then.

**Comment by trvrb** [comment]
I'm thinking that this PR has gone on long enough. It's now effectively `master`. There are some small changes that are likely to happen with `--group-by` and `augur translate`, but these can happen on their own branches. There are additional larger questions about `augur treetime` (#133) and `node_data.json` (#134) that are separate from this PR in it's current state.

Unless someone is opposed, I'm going to merge this at 12:30pm Seattle time.

**Comment by tsibley** [comment]
FWIW, the lassa and ebola builds are broken on this branch since they haven't had their options updated.

**Comment by tsibley** [comment]
It may not be worth updating those until we touch them to move them into per-pathogen repos?

**Comment by trvrb** [comment]
>It may not be worth updating those until we touch them to move them into per-pathogen repos?

This was my thought exactly. Let's get Zika and TB to a state where we're close to 100% happy and then replicate. Otherwise there will be continued duplicated effort.

**Comment by emmahodcroft** [comment]
@tsibley  Yes, of course:

Required for both VCF and fasta:
- `--tree` - used to determine AA changes between nodes (the tree structure is needed)*
- `--reference-sequence` - a GFF or GB file that names and describes the genes (start, end, sense) 
- `--output` - JSON file that contains the AA mutations in the correct format to send to `export` and incorporate into `tree.json` final output file for auspice

Optional but works for both:
- `--genes` - list of genes to translate (as arguments or as a file), if you don't want all of them (particularly relevant for bacteria, less important for viruses)

Fasta-specific:
- `--node-data` - JSON; used just to get the sequences to translate (as it contains both ancestral and original) _(required)_
- `--alignment-output` - writes 1 fasta file per gene, with the AA translation of that gene _(optional)_

VCF-specific:
- `--vcf-input` - provides the sequences to translate (should contain both ancestral and original) _(required)_
- `--vcf-reference` - needed anytime you're reading in a VCF to get non-variable sites _(required)_
- `--vcf-output` - VCF equivalent of `--alignment-output` but in pseudo-VCF format to save space. Only one file is written because you can specify 'CHROM' (chromosome), which I use to distinguish the genes. A fasta file 'reference' translation for each gene is also generated and named based on the `vcf-output` file name _(optional)_

Because VCF users _have_ a `node_data` file, they may think they should pass it in with the matching argument, even though it's not needed as for VCF input ancestral sequences are written to a new VCF, not the JSON. This does no harm as it's never read if they've passed a VCF file with `--vcf-input`, but could be clearer.


*The 'help' on this needs to be updated - as far as I can tell no tree is ever built, the user must provide one.

**Comment by tsibley** [comment]
Thanks @emmahodcroft, that's a very useful breakdown!  We can hopefully use it to clarify arguments.

**Pull request merged**